File: ios/Yomiyasu/Core/Models/LibraryModels.swift
import Foundation

enum Variant: String, Codable, Sendable, CaseIterable {
    case manga
    case novela

    var title: String {
        switch self {
        case .manga: "Manga"
        case .novela: "Novela"
        }
    }

    var staticFolder: String {
        switch self {
        case .manga: "mangas"
        case .novela: "novelas"
        }
    }
}

enum LibraryVariant: String, Sendable, CaseIterable, Identifiable {
    case manga
    case novela
    case all

    var id: String { rawValue }

    var title: String {
        switch self {
        case .manga: "Mangas"
        case .novela: "Novelas"
        case .all: "Todo"
        }
    }
}

enum MainView: String, Codable, Sendable, CaseIterable, Identifiable {
    case manga
    case novels
    case both

    var id: String { rawValue }

    var title: String {
        switch self {
        case .manga: "Solo manga"
        case .novels: "Solo novelas"
        case .both: "Manga y novelas"
        }
    }
}

enum SerieStatus: String, Codable, Sendable, CaseIterable, Identifiable {
    case publishing = "PUBLISHING"
    case ended = "ENDED"

    var id: String { rawValue }

    var title: String {
        switch self {
        case .publishing: "En publicación"
        case .ended: "Finalizada"
        }
    }
}

enum ProgressStatus: String, Codable, Sendable {
    case unread
    case reading
    case completed

    var title: String {
        switch self {
        case .unread: "Sin leer"
        case .reading: "Leyendo"
        case .completed: "Leído"
        }
    }
}

enum ProgressFilter: String, Codable, Sendable, CaseIterable, Identifiable {
    case all
    case unread
    case reading
    case completed

    var id: String { rawValue }

    var title: String {
        switch self {
        case .all: "Todos"
        case .unread: "Sin leer"
        case .reading: "Leyendo"
        case .completed: "Leídos"
        }
    }
}

struct SortValue: Sendable, Equatable, Hashable {
    var key: String
    var descending: Bool

    var rawValue: String {
        descending ? "!\(key)" : key
    }

    static let seriesDefault = SortValue(key: "sortName", descending: false)
    static let seriesNewest = SortValue(key: "_id", descending: true)
    static let seriesRecent = SortValue(key: "lastModifiedDate", descending: true)
    static let seriesValoration = SortValue(key: "valoration", descending: true)
    static let booksDefault = SortValue(key: "sortName", descending: false)
    static let booksNewest = SortValue(key: "_id", descending: true)
}

struct ReadlistEntry: Codable, Sendable, Equatable {
    let id: String?
    let serie: String?
    let addedDate: Date?

    private enum CodingKeys: String, CodingKey {
        case id = "_id"
        case serie
        case addedDate
    }
}

struct ReadlistValue: Decodable, Sendable, Equatable {
    let isInReadlist: Bool

    init(isInReadlist: Bool) {
        self.isInReadlist = isInReadlist
    }

    init(from decoder: any Decoder) throws {
        let container = try decoder.singleValueContainer()

        if let value = try? container.decode(Bool.self) {
            isInReadlist = value
        } else if let entry = try? container.decode(ReadlistEntry.self) {
            isInReadlist = entry.id != nil
        } else {
            isInReadlist = false
        }
    }
}

enum CurrentBook: Decodable, Sendable, Equatable {
    case id(String)
    case book(Book)

    init(from decoder: any Decoder) throws {
        let container = try decoder.singleValueContainer()

        if let id = try? container.decode(String.self) {
            self = .id(id)
            return
        }

        self = .book(try container.decode(Book.self))
    }

    var id: String {
        switch self {
        case .id(let id): id
        case .book(let book): book.id
        }
    }
}

struct Review: Codable, Identifiable, Sendable, Equatable {
    let id: String
    let user: String?
    let serie: String?
    let userLevel: String?
    let difficulty: Int?
    let valoration: Int?
    let comment: String?
    let name: String?

    private enum CodingKeys: String, CodingKey {
        case id = "_id"
        case user
        case serie
        case userLevel
        case difficulty
        case valoration
        case comment
        case name
    }
}

struct Serie: Decodable, Identifiable, Sendable, Equatable {
    let id: String
    let path: String?
    let variant: Variant?
    let visibleName: String
    let sortName: String?
    let bookCount: Int?
    let difficulty: Double?
    let createdDate: Date?
    let lastModifiedDate: Date?
    let status: SerieStatus?
    let summary: String?
    let authors: [String]?
    let genres: [String]?
    let missing: Bool?
    let valoration: Double?
    let alternativeNames: [String]?

    let unreadBooks: Int?
    let paused: Bool?
    let readlist: ReadlistValue?
    let thumbnailPath: String?
    let currentBook: CurrentBook?
    let reviews: [Review]?

    private enum CodingKeys: String, CodingKey {
        case id = "_id"
        case path
        case variant
        case visibleName
        case sortName
        case bookCount
        case difficulty
        case createdDate
        case lastModifiedDate
        case status
        case summary
        case authors
        case genres
        case missing
        case valoration
        case alternativeNames
        case unreadBooks
        case paused
        case readlist
        case thumbnailPath
        case currentBook
        case reviews
    }
}

extension Serie {
    var isInReadlist: Bool { readlist?.isInReadlist ?? false }
    var unreadCount: Int { unreadBooks ?? 0 }
    var totalBooks: Int { bookCount ?? 0 }
    var isPaused: Bool { paused ?? false }
    var displayGenres: [String] { genres ?? [] }
    var displayAuthors: [String] { authors ?? [] }
    var progressFraction: Double {
        guard totalBooks > 0 else { return 0 }
        return Double(totalBooks - unreadCount) / Double(totalBooks)
    }
    var plainSummary: String {
        (summary ?? "")
            .replacingOccurrences(of: "<[^>]+>", with: "", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

/// Formato del tomo: mokuro (html con OCR) o carpeta de imágenes sin html.
/// El backend no manda el campo en libros antiguos: ausente = mokuro.
enum BookFormat: String, Decodable, Sendable {
    case mokuro
    case images
}

struct Book: Decodable, Identifiable, Sendable, Equatable {
    let id: String
    let path: String?
    let serie: String?
    let seriePath: String?
    let pages: Int?
    let visibleName: String
    let sortName: String?
    let imagesFolder: String?
    let thumbnailPath: String?
    let createdDate: Date?
    let releaseDate: Date?
    let lastModifiedDate: Date?
    let missing: Bool?
    let characters: Int?
    let pageChars: [Int]?
    let variant: Variant?
    let mokured: Bool?
    let format: BookFormat?
    /// Nombres de las imágenes del tomo (solo tomos "images").
    let pagePaths: [String]?

    let status: ProgressStatus?
    let lastProgress: ReadProgress?
    let readlist: ReadlistValue?
    let type: String?

    private enum CodingKeys: String, CodingKey {
        case id = "_id"
        case path
        case serie
        case seriePath
        case pages
        case visibleName
        case sortName
        case imagesFolder
        case thumbnailPath
        case createdDate
        case releaseDate
        case lastModifiedDate
        case missing
        case characters
        case pageChars
        case variant
        case mokured
        case format
        case pagePaths
        case status
        case lastProgress
        case readlist
        case type
    }
}

extension Book {
    var resolvedStatus: ProgressStatus { status ?? .unread }
    var isMokured: Bool { mokured ?? false }
    var isImageFolder: Bool { format == .images }
    var progressFraction: Double {
        guard let pages, pages > 0, let currentPage = lastProgress?.currentPage else {
            if let characters, characters > 0, let readCharacters = lastProgress?.characters {
                return min(1, Double(readCharacters) / Double(characters))
            }
            return 0
        }
        return min(1, Double(currentPage) / Double(pages))
    }
}

struct ReadProgress: Codable, Identifiable, Sendable, Equatable {
    let id: String?
    let book: String?
    let serie: String?
    let startDate: Date?
    let lastUpdateDate: Date?
    let endDate: Date?
    let time: Int?
    let currentPage: Int?
    let status: ProgressStatus?
    let paused: Bool?
    let characters: Int?
    let variant: Variant?

    private enum CodingKeys: String, CodingKey {
        case id = "_id"
        case book
        case serie
        case startDate
        case lastUpdateDate
        case endDate
        case time
        case currentPage
        case status
        case paused
        case characters
        case variant
    }
}

struct SeriesPage: Decodable, Sendable {
    let data: [Serie]
    let pages: Int
}

struct AlphabetGroup: Decodable, Sendable, Identifiable, Equatable {
    let group: String
    let count: Int

    var id: String { group }

    var displayName: String {
        switch group {
        case "all": "Todo"
        case "#": "#"
        default: group.uppercased()
        }
    }
}

struct GenresAndArtists: Decodable, Sendable, Equatable {
    let genres: [String]
    let authors: [String]
}

struct ReadlistRequest: Encodable, Sendable {
    let serie: String
}

struct EmptyResponse: Decodable, Sendable {
    init(from decoder: any Decoder) throws {}
}

File: android/app/src/main/java/es/manabe/yomiyasu/core/models/LibraryModels.kt
package es.manabe.yomiyasu.core.models

import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.jsonObject

@Serializable
enum class Variant {
    @SerialName("manga")
    Manga,

    @SerialName("novela")
    Novela,
    ;

    val title: String
        get() = when (this) {
            Manga -> "Manga"
            Novela -> "Novela"
        }

    val staticFolder: String
        get() = when (this) {
            Manga -> "mangas"
            Novela -> "novelas"
        }
}

@Serializable
enum class LibraryVariant(val rawValue: String) {
    @SerialName("manga")
    Manga("manga"),

    @SerialName("novela")
    Novela("novela"),

    @SerialName("all")
    All("all"),
    ;

    val title: String
        get() = when (this) {
            Manga -> "Mangas"
            Novela -> "Novelas"
            All -> "Todo"
        }
}

@Serializable
enum class MainView(val rawValue: String) {
    @SerialName("manga")
    Manga("manga"),

    @SerialName("novels")
    Novels("novels"),

    @SerialName("both")
    Both("both"),
    ;

    val title: String
        get() = when (this) {
            Manga -> "Solo manga"
            Novels -> "Solo novelas"
            Both -> "Manga y novelas"
        }
}

@Serializable
enum class SerieStatus(val rawValue: String) {
    @SerialName("PUBLISHING")
    Publishing("PUBLISHING"),

    @SerialName("ENDED")
    Ended("ENDED"),
    ;

    val title: String
        get() = when (this) {
            Publishing -> "En publicación"
            Ended -> "Finalizada"
        }
}

@Serializable
enum class ProgressStatus {
    @SerialName("unread")
    Unread,

    @SerialName("reading")
    Reading,

    @SerialName("completed")
    Completed,
    ;

    val title: String
        get() = when (this) {
            Unread -> "Sin leer"
            Reading -> "Leyendo"
            Completed -> "Leído"
        }

    val key: String
        get() = when (this) {
            Unread -> "unread"
            Reading -> "reading"
            Completed -> "completed"
        }
}

enum class ProgressFilter(val rawValue: String?) {
    All(null),
    Unread("unread"),
    Reading("reading"),
    Completed("completed"),
    ;

    val title: String
        get() = when (this) {
            All -> "Todos"
            Unread -> "Sin leer"
            Reading -> "Leyendo"
            Completed -> "Leídos"
        }
}

data class SortValue(val key: String, val descending: Boolean) {
    val rawValue: String
        get() = if (descending) "!$key" else key

    companion object {
        val SeriesDefault = SortValue("sortName", false)
        val SeriesNewest = SortValue("_id", true)
        val SeriesRecent = SortValue("lastModifiedDate", true)
        val BooksDefault = SortValue("sortName", false)
        val BooksNewest = SortValue("_id", true)
    }
}

@Serializable
data class ReadlistEntry(
    @SerialName("_id") val id: String? = null,
    val serie: String? = null,
    val addedDate: String? = null,
)

object ReadlistValueSerializer : KSerializer<ReadlistValue> {
    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("ReadlistValue")

    override fun deserialize(decoder: Decoder): ReadlistValue {
        val jsonDecoder = decoder as? JsonDecoder
            ?: return ReadlistValue(false)
        val element: JsonElement = jsonDecoder.decodeJsonElement()

        return when (element) {
            is JsonPrimitive -> ReadlistValue(element.booleanOrNull ?: false)
            is JsonObject -> {
                val id = element["_id"]
                ReadlistValue(id != null && id !is JsonNull)
            }
            else -> ReadlistValue(false)
        }
    }

    override fun serialize(encoder: Encoder, value: ReadlistValue) {
        encoder.encodeBoolean(value.isInReadlist)
    }
}

@Serializable(with = ReadlistValueSerializer::class)
data class ReadlistValue(val isInReadlist: Boolean)

object CurrentBookSerializer : KSerializer<CurrentBook> {
    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("CurrentBook")

    override fun deserialize(decoder: Decoder): CurrentBook {
        val jsonDecoder = decoder as? JsonDecoder
            ?: throw IllegalArgumentException("CurrentBook requiere un decoder JSON")
        val element: JsonElement = jsonDecoder.decodeJsonElement()

        return when (element) {
            is JsonPrimitive -> CurrentBook.Id(element.content)
            is JsonObject -> CurrentBook.BookValue(
                jsonDecoder.json.decodeFromJsonElement(Book.serializer(), element),
            )
            else -> throw IllegalArgumentException("CurrentBook inválido")
        }
    }

    override fun serialize(encoder: Encoder, value: CurrentBook) {
        when (value) {
            is CurrentBook.Id -> encoder.encodeString(value.id)
            is CurrentBook.BookValue -> encoder.encodeSerializableValue(Book.serializer(), value.book)
        }
    }
}

@Serializable(with = CurrentBookSerializer::class)
sealed class CurrentBook {
    abstract val id: String

    data class Id(override val id: String) : CurrentBook()
    data class BookValue(val book: Book) : CurrentBook() {
        override val id: String get() = book.id
    }
}

@Serializable
data class Review(
    @SerialName("_id") val id: String,
    val user: String? = null,
    val serie: String? = null,
    val userLevel: String? = null,
    val difficulty: Int? = null,
    val valoration: Int? = null,
    val comment: String? = null,
    val name: String? = null,
)

@Serializable
data class Serie(
    @SerialName("_id") val id: String,
    val path: String? = null,
    val variant: Variant? = null,
    val visibleName: String = "",
    val sortName: String? = null,
    val bookCount: Int? = null,
    val difficulty: Double? = null,
    val createdDate: String? = null,
    val lastModifiedDate: String? = null,
    val status: SerieStatus? = null,
    val summary: String? = null,
    val authors: List<String>? = null,
    val genres: List<String>? = null,
    val missing: Boolean? = null,
    val valoration: Double? = null,
    val alternativeNames: List<String>? = null,
    val unreadBooks: Int? = null,
    val paused: Boolean? = null,
    val readlist: ReadlistValue? = null,
    val thumbnailPath: String? = null,
    val currentBook: CurrentBook? = null,
    val reviews: List<Review>? = null,
) {
    val isInReadlist: Boolean get() = readlist?.isInReadlist ?: false
    val unreadCount: Int get() = unreadBooks ?: 0
    val totalBooks: Int get() = bookCount ?: 0
    val isPaused: Boolean get() = paused ?: false
    val displayGenres: List<String> get() = genres ?: emptyList()
    val displayAuthors: List<String> get() = authors ?: emptyList()

    val progressFraction: Double
        get() = if (totalBooks > 0) {
            (totalBooks - unreadCount).toDouble() / totalBooks.toDouble()
        } else {
            0.0
        }

    val plainSummary: String
        get() = (summary ?: "")
            .replace(Regex("<[^>]+>"), "")
            .trim()
}

@Serializable
data class Book(
    @SerialName("_id") val id: String,
    val path: String? = null,
    val serie: String? = null,
    val seriePath: String? = null,
    val pages: Int? = null,
    val visibleName: String = "",
    val sortName: String? = null,
    val imagesFolder: String? = null,
    val thumbnailPath: String? = null,
    val createdDate: String? = null,
    val releaseDate: String? = null,
    val lastModifiedDate: String? = null,
    val missing: Boolean? = null,
    val characters: Int? = null,
    val pageChars: List<Int>? = null,
    val variant: Variant? = null,
    val mokured: Boolean? = null,
    val format: String? = null,
    val pagePaths: List<String>? = null,
    val status: ProgressStatus? = null,
    val lastProgress: ReadProgress? = null,
    val readlist: ReadlistValue? = null,
    val type: String? = null,
) {
    val resolvedStatus: ProgressStatus get() = status ?: ProgressStatus.Unread
    val isMokured: Boolean get() = mokured ?: false

    /** Tomo de solo imágenes, sin html de mokuro. Ausente = mokuro. */
    val isImageFolder: Boolean get() = format == "images"

    val progressFraction: Double
        get() {
            val pageCount = pages
            val currentPage = lastProgress?.currentPage
            if (pageCount != null && pageCount > 0 && currentPage != null) {
                return minOf(1.0, currentPage.toDouble() / pageCount.toDouble())
            }
            val totalChars = characters
            val readCharacters = lastProgress?.characters
            if (totalChars != null && totalChars > 0 && readCharacters != null) {
                return minOf(1.0, readCharacters.toDouble() / totalChars.toDouble())
            }
            return 0.0
        }
}

@Serializable
data class ReadProgress(
    @SerialName("_id") val id: String? = null,
    val book: String? = null,
    val serie: String? = null,
    val startDate: String? = null,
    val lastUpdateDate: String? = null,
    val endDate: String? = null,
    val time: Int? = null,
    val currentPage: Int? = null,
    val status: ProgressStatus? = null,
    val paused: Boolean? = null,
    val characters: Int? = null,
    val variant: Variant? = null,
)

@Serializable
data class SeriesPage(
    val data: List<Serie> = emptyList(),
    val pages: Int = 1,
)

@Serializable
data class AlphabetGroup(
    val group: String,
    val count: Int = 0,
) {
    val displayName: String
        get() = when (group) {
            "all" -> "Todo"
            "#" -> "#"
            else -> group.uppercase()
        }
}

@Serializable
data class GenresAndArtists(
    val genres: List<String> = emptyList(),
    val authors: List<String> = emptyList(),
)

@Serializable
data class ReadlistRequest(val serie: String)

@Serializable
class EmptyResponse
